### ISSUE TITLE
AMPLIFY_DIFF_BACKEND does not work

### ISSUE THREAD

**Issue by Voyager-Two** [aws-amplify/amplify-hosting#2921]
### Before opening, please confirm:

- [X] I have checked to see if my question is addressed in the [FAQ](https://github.com/aws-amplify/amplify-hosting/blob/master/FAQ.md).
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-hosting/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-hosting/blob/master/CONTRIBUTING.md).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### Region

us-east-1

### Amplify Hosting feature

Backend builds

### Describe the bug

Docs: https://docs.aws.amazon.com/amplify/latest/userguide/build-settings.html#enable-diff-backend

Backend builds even with:
- `AMPLIFY_DIFF_BACKEND` set to `true`.
- No code changes to `amplify` directory.
- With no custom `backend` commands in `amplify.yml` (apart from `amplifyPush --simple`)

**This often adds 4 min unnecessary overhead to builds.**

![image](https://user-images.githubusercontent.com/3676208/183966280-e5e6c928-5d71-4f00-ac00-cfe8ef33fe53.png)

[In this comment](https://github.com/aws-amplify/amplify-hosting/issues/2385#issuecomment-1006392315) it is stated we can use `amplifyPush --simple` command, and it should still work.

But it does not:
![image](https://user-images.githubusercontent.com/3676208/183966194-8c9e737a-297b-4d0d-bfc4-53255d0eed8d.png)

![image](https://user-images.githubusercontent.com/3676208/183967263-8f9aa4bb-45a5-4e96-a764-67126c8d035d.png)

---

[Docs](https://docs.aws.amazon.com/amplify/latest/userguide/build-settings.html#enable-diff-backend) states the following:
> If you currently have custom commands specified in the build settings of your backend phase, conditional backend builds won't work.

However, we require `aws-exports` as it's deployed to a different environment. So we need to use `amplifyPush --simple`.

### Expected behavior

Skip backend builds if no changes to `amplify` dir, without having to give up `aws-exports` functionality.

### Reproduction steps

See above

### Build Settings

```yaml
version: 1
backend:
  phases:
    build:
      commands:
        - amplifyPush --simple
frontend:
  phases:
    preBuild:
      commands:
        - yarn install
    build:
...
```

### Additional information

A simple workaround could be to add `[skip backend]` to commit messages option. Just like how we can add `[skip-cd]`.

**Comment by hloriana** [comment]
Hi @Voyager-Two I'm able to successfully use the `AMPLIFY_DIFF_BACKEND` variable and see that it is working. Can you please provide your app ID so we can further investigate the behavior?

**Comment by Voyager-Two** [comment]
@hloriana I don't feel comfortable providing the App ID in a public forum. What does your `amplify.yml` look like while testing?

**Comment by hloriana** [comment]
Hi @Voyager-Two it looks like you set the environment variable correctly but Amplify is somehow not detecting it. Unfortunately I still haven't been able to reproduce this behavior even with just running `amplifyPush --simple`. 

Can you: 
- confirm if  this is happening in multiple apps or just this one app
- try setting the `AMPLIFY_SKIP_BACKEND_BUILD` to `true` and confirm if your backend builds can be skipped with this variable instead

**Comment by oste** [comment]
I have a related question. When using `AMPLIFY_DIFF_BACKEND` can we still use `amplifyPush --simple` as part of the backend build? I get the sense that there can't be any commands whatsoever in the build phase. I suppose it would help if `custom commands` can be defined more clearly here states https://docs.aws.amazon.com/amplify/latest/userguide/build-settings.html

where the docs state:

> If you currently have custom commands specified in the build settings of your backend phase, conditional backend builds won't work. If you want those custom commands to run, you must move them to the frontend phase of your build settings in your app's amplify.yml file.

**Comment by oste** [comment]
I should follow up, if we can't use amplifyPush --simple as part of the build command with AMPLIFY_DIFF_BACKEND how should we go about deploying backend resources conditionally?

**Comment by Voyager-Two** [comment]
@oste Yes you should be able to. But there seems to be a bug where it always does backend build even if there are no changes in `amplify` directory.

**Comment by Engineersticity** [comment]
If anyone is still struggling with the native AMPLIFY_DIFF_BACKEND failing when using custom commands, I ended up writing a lightweight NPM package to handle this cleanly at the CI/CD execution level instead of relying on Amplify's environment variables.

You can wrap whatever custom script you want with `npx skip-amplify-backend <your custom command>`.

It validates the git history directly in the build container. If nothing changed in the backend folder, it skips your custom deploy command, but it will automatically run `amplify pull` to pull the latest `aws-exports.js` so your build doesn't break in the next phase.

Repo and instructions are here: https://www.npmjs.com/package/skip-amplify-backend
